src/compiler/checker.ts

```diff
@@ -23563,12 +23565,40 @@ namespace ts {
             // and the type of the node includes type variables with constraints that are nullable, we fetch the
             // apparent type of the node *before* performing control flow analysis such that narrowings apply to
             // the constraint type.
-            if (type && isConstraintPosition(node) && forEachType(type, typeHasNullableConstraint)) {
+            if (type && isConstraintPosition(node) && someType(type, typeHasNullableConstraint)) {
                 return mapType(getWidenedType(type), getBaseConstraintOrType);
             }
             return type;
         }
 
+        function isTypeVariableWithUnionConstraint(type: Type) {
+            return !!(type.flags & TypeFlags.Instantiable && getBaseConstraintOrType(type).flags & TypeFlags.Union);
+        }
+
+        function containsTypeVariable(type: Type): boolean {
+            return !!(type.flags & TypeFlags.Instantiable || type.flags & TypeFlags.UnionOrIntersection && some((<UnionOrIntersectionType>type).types, containsTypeVariable));
+        }
+
+        function hasContextualTypeWithNoTypeVariables(node: Expression) {
+            const contextualType = !((isJsxOpeningElement(node.parent) || isJsxSelfClosingElement(node.parent)) && node.parent.tagName === node) && getContextualType(node);
+            return contextualType && !someType(contextualType, containsTypeVariable);

```

Comment:
Asking for the contextual type sometimes leads to circularities, and I think this function is called in a couple of places that wouldn't otherwise do that.

Uh...I'm not sure how to figure out if this is a problem, it's just worth thinking about.

Reply:
Yup, I had to exclude JSX elements for exactly that reason. I added a comment to that effect.